conformance/src/main/scala/za/co/absa/enceladus/conformance/interpreter/DynamicInterpreter.scala

```diff
@@ -62,123 +62,198 @@ object DynamicInterpreter {
 
     // Fold left on rules
     var rulesApplied = 0
-    val conformedDf = steps.foldLeft(dfInputWithIdForWorkaround)({
-      case (df, rule) =>
-        val (ruleAppliedDf, ec) = applyConformanceRule(df, rule, explodeContext)
+    val conformedDf = interpreters.foldLeft(dfInputWithIdForWorkaround)({
+      case (df, interpreter) =>
+        val explosionStateCopy = new ExplosionState(explosionState.explodeContext)
+        val ruleAppliedDf = interpreter.conform(df)(spark, explosionStateCopy, dao, progArgs)
 
-        val conformedDf = if (explodeContext.explosions.isEmpty &&
+        val conformedDf = if (explosionState.isNoExplosionsApplied &&
           optimizerTimeTracker.isCatalystWorkaroundRequired(ruleAppliedDf, rulesApplied)) {
           // Apply a workaround BEFORE applying the rule so that the execution plan generation still runs fast
-          val (workAroundDf, ecInner) = applyConformanceRule(
-            optimizerTimeTracker.applyCatalystWorkaround(df),
-            rule,
-            explodeContext)
-          explodeContext = ecInner
+          val workAroundDf = interpreter.conform(optimizerTimeTracker.applyCatalystWorkaround(df))
           optimizerTimeTracker.recordExecutionPlanOptimizationTime(workAroundDf)
           workAroundDf
         } else {
-          explodeContext = ec
+          explosionState.explodeContext = explosionStateCopy.explodeContext
           ruleAppliedDf
         }
         rulesApplied += 1
-        applyRuleCheckpoint(rule, conformedDf, progArgs.persistStorageLevel, explodeContext)
+        interpreter.conformanceRule match {
+          case Some(rule) => applyRuleCheckpoint(rule, conformedDf, progArgs.persistStorageLevel, explosionState.explodeContext)
+          case None       => conformedDf
+        }
     })
     optimizerTimeTracker.cleanupWorkaroundDf(conformedDf)
   }
 
   /**
-    * Gets the list of conformance rules from the context and applies optimization rearrangements if needed
+    * Transforms a list of conformance rules to a list of conformance rule interpreters.
+    * For most conformance rules there is only one interpreter to apply. But mapping rule
+    * has several strategies. Optimizer chooses which strategy to use and provides an
+    * interpreter for each strategy.
     *
-    * @return A list of conformance rules
+    * @param rules  A list of conformance rules.
+    * @param schema A schema of a DataFrame to be conformed.
+    * @return A list of conformance rule interpreters.
     */
-  def getConformanceSteps(implicit ictx: InterpreterContext): List[ConformanceRule] = {
-    val steps = ictx.conformance.conformance.sortBy(_.order)
+  def getInterpreters(rules: List[ConformanceRule], schema: StructType)
+                     (implicit ictx: InterpreterContext): List[RuleInterpreter] = {
+
+    val groupedRules = groupMappingRules(rules, schema)
+
+    // Get unoptimized list of interpreters
```

Comment:
This comment states the exact opposite of the code on the next line. I'd just remove it.